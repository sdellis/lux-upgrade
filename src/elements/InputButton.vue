<template>
  <button :type="!!type ? type : false"
    :class="['lux-button', variation, size, {'lux-expanded' : (block==true) }]"
    :disabled="disabled"
    @click="buttonClicked($event)">
    <slot>
    {{ label }}
    </slot>
    <span v-if="variation === 'dropdown'" class="lux-dropdown-indicator"> &#9660;</span>
  </button>
</template>

<script>
/**
 * Buttons are used to toggle something in the interface or trigger new
 * content in the same context.
 */
export default {
  name: "InputButton",
  status: "ready",
  release: "1.0.0",
  type: "Element",
  data: function() {
    return {
      label: "Submit",
    }
  },
  props: {
    /**
     * The button's variations `solid, outline, text, dropdown`
     */
    variation: {
      type: String,
      default: "button",
      validator: value => {
        return value.match(/(solid|outline|text|dropdown)/)
      },
    },
    /**
     * The button's type attribute `button, submit`
     */
    type: {
      type: String,
      default: "",
      validator: value => {
        return value.match(/(|button|submit)/)
      },
    },
    /**
     * Sets the size of the button `small, medium, large`
     */
    size: {
      type: String,
      default: "medium",
      validator: value => {
        return value.match(/(small|medium|large)/)
      },
    },
    /**
     * Whether the button extends the full available width or not
     */
    block: {
      type: Boolean,
      default: false,
    },
    /**
     * Whether the button is disabled or not
     * `true, false`
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * Clicking this button can emit a custom event that should trigger an alert.
     * You must supply an alertStatus and alertMessage, like so:
     * { 'alertStatus': 'success', 'alertMessage': 'This is my message.'}
     */
    customAlertEvent: {
      type: Object,
      default: null,
    },
  },
  methods: {
    buttonClicked(value) {
      if (this.customAlertEvent) {
        this.$emit("system-alert", {
          event: value,
          alertStatus: this.customAlertEvent.alertStatus,
          alertMessage: this.customAlertEvent.alertMessage,
        })
      }
      this.$emit("button-clicked", value)
    },
  },
}
</script>

<style lang="scss" scoped>
.lux-button {
  @include reset;
  @include stack-space($space-base);
  font-family: $font-family-text;
  font-size: $font-size-base;
  line-height: $line-height-base;
  border: 0;
  border-radius: $border-radius-default;
  color: $color-rich-black;
  cursor: pointer;
  padding: 0.5rem 1rem;
  margin: 0.25rem;
  text-decoration: none;
  text-align: center;
  transition: background 250ms ease-in-out, transform 150ms ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  &:active {
    transform: scale(0.99);
  }

  &.solid {
    background: $color-bleu-de-france;
    color: $color-white;
    &:hover,
    &:focus {
      background: $color-bleu-de-france-darker;
    }
  }

  &.text {
    background-color: transparent;
  }

  &.dropdown {
    background: $color-white;
    box-shadow: $box-shadow-small;
    color: $color-rich-black;

    .lux-dropdown-indicator {
      line-height: 1;
      font-size: $font-size-x-small;
      color: $color-grayscale-darker;
      margin-left: $space-xx-small;
    }
  }
}

.small {
  font-size: $font-size-small;

  & .lux-dropdown-indicator {
    font-size: calc(#{$font-size-small} - 2px);
  }
}

.medium {
  font-size: $font-size-base;

  & .lux-dropdown-indicator {
    font-size: calc(#{$font-size-base} - 3px);
  }
}

.large {
  font-size: $font-size-large;

  & .lux-dropdown-indicator {
    font-size: calc(#{$font-size-large} - 4px);
  }
}

.lux-expanded {
  display: block;
  width: 100%;
  max-width: 100%;
}

[disabled] {
  background: $color-grayscale-light;
  cursor: not-allowed;
  &:hover,
  &:focus {
    background: $color-grayscale-light;
  }
}
</style>

<docs>
  ```jsx
  <div>
    <input-button variation="solid" size="small">Apply Changes</input-button>
    <input-button type="button" variation="solid">Apply Changes</input-button>
    <input-button type="button" variation="solid" size="large" disabled>Apply Changes</input-button>

    <input-button type="button" variation="solid" block></input-button>

    <input-button type="button" variation="text">Manage Files</input-button>
  </div>
  ```
</docs>
