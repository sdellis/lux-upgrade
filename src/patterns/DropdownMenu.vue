<template>
  <component :is="element" :class="['lux-dropdown-menu']">
    <input-button v-on:button-clicked="buttonClicked($event)" class="lux-dropdown-button" variation="dropdown">{{ buttonLabel }}</input-button>
    <menu-bar v-bind:class="{ 'lux-show': isVisible }" v-on:menu-item-clicked="menuItemClicked($event)" :type="type" active="All" :menu-items="menuItems"/>
  </component>
</template>

<script>
/**
 * Dropdowns allows a user to select a value from a series of options.
 */
export default {
  name: "DropdownMenu",
  status: "ready",
  release: "1.0.0",
  type: "Pattern",
  data: function() {
    return {
      isVisible: false,
    }
  },
  props: {
    /**
     * The html element name used for the container
     */
    element: {
      type: String,
      default: "div",
    },
    /**
     * The menu items in the dropdown
     */
    buttonLabel: {
      type: String,
      default: "Dropdown",
    },
    /**
     * Whether the dropdown includes links or buttons as menu items
     */
    type: {
      type: String,
      default: "buttons",
      validator: value => {
        return value.match(/(links|buttons)/)
      },
    },
    /**
     * The html element name used for the container
     */
    menuItems: {
      type: Array,
      default: "div",
    },
    /**
     * Style variation to give additional meaning.
     * `default, disabled, active`
     */
    align: {
      type: String,
      default: "left",
      validator: value => {
        return value.match(/(left|right)/)
      },
    },
  },
  methods: {
    buttonClicked(value) {
      this.isVisible = !this.isVisible
      this.$emit("button-clicked", value)
    },
    menuItemClicked(value) {
      this.isVisible = false
      this.$emit("menu-item-clicked", value)
    },
  },
}
</script>

<style lang="scss" scoped>
.lux-dropdown-menu {
  @include reset;
  @include stack-space($space-small);
  margin: 0.25rem;
  font-family: $font-family-text;
  font-size: $font-size-base;
  line-height: 1;
  color: set-text-color($color-rich-black, $color-white);
  position: relative;
  display: inline-block;
  top: 0;

  .lux-dropdown {
    position: relative;
    vertical-align: middle;
    line-height: 1;
    margin: 0;

    &:hover,
    &:focus {
      background: $color-grayscale-lighter;
    }
  }

  .lux-menu,
  .lux-nav {
    opacity: 0;
    visibility: hidden;
  }

  .lux-show {
    opacity: 1;
    z-index: 100000;
    visibility: visible;
  }

  /deep/ .lux-nav {
    background: $color-white;
    box-shadow: $box-shadow-small;

    position: absolute;
    min-width: calc(100% - 1px);
    margin: 1px 0 0;
    z-index: 100000;

    ul {
      list-style-type: none;
      margin: 0;
      padding: 0;
    }

    li {
      display: block;
    }

    .lux-nav-item {
      background: $color-white;
      border: 0;
      color: $color-rich-black;
      cursor: pointer;
      display: block;
      padding: 0.5rem 1rem;
      margin: 0;
      width: 100%;
      text-align: left;
      font-size: $font-size-small;

      &:hover,
      &:focus {
        color: $color-rich-black;
        background: $color-grayscale-lighter;
        transition: opacity 0.1s ease;
      }

      &:active {
        transform: scale(0.99);
      }
    }

    .lux-disabled {
      color: $color-grayscale-dark;

      &:hover,
      &:focus {
        background: $color-white;
        cursor: not-allowed;
        outline: none;
      }
    }
  }
}
</style>

<docs>
  ```jsx
  <dropdown-menu button-label="Select Options" :menu-items="[
    {name: 'All', component: 'All'},
    {name: 'None', component: 'None'},
    {name: 'Alternate', component: 'Alternate', disabled: true},
    {name: 'Inverse', component: 'Inverse'}
  ]"/>
  ```
</docs>
