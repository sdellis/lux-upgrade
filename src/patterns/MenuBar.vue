<template>
  <nav v-if="type === 'links'" class="lux-nav">
    <ul>
      <li v-for="(item, index) in menuItems">
        <a
          :key="index"
          :href="item.href"
          :class="[
            'lux-nav-item',
            {'lux-active': localActive === item.component}]"
          v-html="item.name"
          @click="menuItemClicked($event)">
        </a>
      </li>
    </ul>
  </nav>

  <div v-else-if="type === 'buttons'" class="lux-menu">
    <ul>
      <li v-for="(item, index) in menuItems">
        <button
          :key="index"
          :href="item.href"
          :class="[
            'lux-menu-item',
            {'lux-active': localActive === item.component},
            {'lux-disabled': item.disabled}]"
          v-html="item.name"
          :disabled="item.disabled"
          @click="menuItemClicked($event, item)">
        </button>
      </li>
    </ul>
  </div>
</template>

<script>
/**
 * Used as main page navigation in templates.
 */
export default {
  name: "MenuBar",
  status: "ready",
  release: "1.0.0",
  type: "Pattern",
  model: {
    prop: "active",
  },
  props: {
    /**
     * The html element types used for the nav bar.
     */
    type: {
      type: String,
      default: "links",
      validator: value => {
        return value.match(/(links|buttons)/)
      },
    },
    /**
     * State which tab is active when initiated (using name of the component).
     */
    active: {
      required: false,
      type: String,
    },
    /**
     * Menu items are options to be displayed to the user.
     */
    menuItems: {
      required: true,
      type: Array,
    },
  },
  computed: {
    localActive: {
      get() {
        return this.active
      },
      set(val) {
        this.$emit("input", val)
      },
    },
  },
  methods: {
    menuItemClicked(value, item) {
      this.$emit("menu-item-clicked", value)
    },
  },
}
</script>

<style lang="scss" scoped>
// Design Tokens with local scope
$color-nav-link: $color-bleu-de-france;
$color-nav-link-active: $color-bleu-de-france;

.lux-nav {
  @include stack-space($space-base);
  font-family: $font-family-text;
  font-size: $font-size-small;
  line-height: $line-height-base;
  color: $color-white;
  text-align: center;
  width: 100%;
  @media #{$media-query-large} {
    // This is how youâ€™d use design tokens with media queries
  }

  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  li {
    display: inline-block;
  }

  a {
    color: $color-nav-link;
    padding: $space-x-small 0;
    margin: 0 $space-x-small;
    text-decoration: none;
    display: inline-block;
    &:hover {
      color: $color-nav-link-active;
    }
    &.lux-active {
      border-bottom: 2px solid $color-nav-link;
      font-weight: $font-weight-bold;
      color: $color-nav-link;
    }
  }
}

.lux-menu {
  background: $color-white;
  box-shadow: $box-shadow-small;

  position: absolute;
  min-width: calc(100% - 1px);
  margin: 1px 0 0;
  z-index: 999;
  transition: opacity 0.1s ease;

  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  li {
    display: block;
  }

  .lux-menu-item {
    background: $color-white;
    border: 0;
    cursor: pointer;
    display: block;
    padding: 0.5rem 1rem;
    width: 100%;
    text-align: left;
    font-size: $font-size-small;

    &:hover,
    &:focus {
      background: $color-grayscale-lighter;
      transition: opacity 0.1s ease;
    }

    &:active {
      transform: scale(0.99);
    }
  }

  .lux-disabled {
    color: $color-grayscale-dark;

    &:hover,
    &:focus {
      background: $color-white;
      cursor: not-allowed;
      outline: none;
    }
  }
}
</style>

<docs>
  ```jsx
  <menu-bar type="links" active="Dashboard" :menu-items="[
    {name: 'Dashboard', component: 'Dashboard', href: '/example/'},
    {name: 'Posts', component: 'Posts', href: '/example/'},
    {name: 'Users', component: 'Users', href: '/example/'},
    {name: 'Settings', component: 'Settings', href: '/example/'}
  ]"/>
  ```
</docs>
