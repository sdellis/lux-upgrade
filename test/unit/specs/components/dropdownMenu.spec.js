import { createLocalVue, mount } from "@vue/test-utils"
import DropdownMenu from "@/patterns/DropdownMenu.vue"
import InputButton from "@/elements/InputButton.vue"
import MenuBar from "@/patterns/MenuBar.vue"

// create an extended `Vue` constructor
const localVue = createLocalVue()

let wrapper

describe("DropdownMenu.vue", () => {
  beforeEach(() => {
    wrapper = mount(DropdownMenu, {
      localVue,
      props: {
        buttonLabel: "Dropdown",
        menuItems: [
          { name: "All", component: "All" },
          { name: "None", component: "None" },
          { name: "Alternate", component: "Alternate", disabled: true },
          { name: "Inverse", component: "Inverse" },
        ],
      },
      stubs: ["input-button", "menu-bar"],
    })
  })

  it("should be of the element that is passed in as a prop", () => {
    expect(wrapper.find("div").exists()).toBe(true)
    expect(wrapper.find("form").exists()).toBe(false)
    // can't setProps here because the component must be regenerated to change its wrapper
    const wrapper2 = mount(DropdownMenu, {
      localVue,
      propsData: {
        element: "form",
        buttonLabel: "Dropdown",
        menuItems: [
          { name: "All", component: "All" },
          { name: "None", component: "None" },
          { name: "Alternate", component: "Alternate", disabled: true },
          { name: "Inverse", component: "Inverse" },
        ],
      },
      stubs: ["input-button", "menu-bar"],
    })
    expect(wrapper2.find("div").exists()).toBe(false)
    expect(wrapper2.find("form").exists()).toBe(true)
  })

  it("should hide menu on button click", () => {
    expect(wrapper.vm.isVisible).toBe(false)
    wrapper.vm.buttonClicked()
    expect(wrapper.vm.isVisible).toBe(true)
  })

  it("should hide menu on menu item click", () => {
    wrapper.vm.buttonClicked()
    expect(wrapper.vm.isVisible).toBe(true)
    wrapper.vm.menuItemClicked()
    expect(wrapper.vm.isVisible).toBe(false)
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
