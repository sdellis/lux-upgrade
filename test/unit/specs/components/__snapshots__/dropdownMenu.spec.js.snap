// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu.vue has the expected html structure 1`] = `
<div
  class="lux-dropdown-menu"
>
  <input-button-stub
    class="lux-dropdown-button"
    variation="dropdown"
  >
    
  </input-button-stub>
   
  <menu-bar-stub
    active="All"
    class=""
    type="buttons"
  />
</div>
`;
